packages/react-dom-bindings/src/client/ReactDOMHostConfig.js

```diff
@@ -3010,3 +3046,257 @@ export function isHostHoistableType(
+      // errored however which we still do not yet have a means to detect. For now
+      // we assume it is loaded.
+      const maybeLoadingState: ?Promise<mixed> = (instance: any)._p;
+      if (
+        maybeLoadingState !== null &&
+        typeof maybeLoadingState === 'object' &&
+        // $FlowFixMe[method-unbinding]
+        typeof maybeLoadingState.then === 'function'
+      ) {
+        const loadingState = maybeLoadingState;
+        state.count++;
+        const ping = onUnsuspend.bind(state);
+        loadingState.then(ping, ping);
+      }
+      resource.instance = instance;
+      markNodeAsHoistable(instance);
+      return;
+    }
+
+    const ownerDocument = getDocumentFromRoot(hoistableRoot);
+
+    const stylesheetProps = stylesheetPropsFromRawProps(props);
+    const preloadProps = preloadPropsMap.get(key);
+    if (preloadProps) {
+      adoptPreloadPropsForStylesheet(stylesheetProps, preloadProps);
+    }
+
+    // Construct and insert a new instance
+    instance = ownerDocument.createElement('link');
+    markNodeAsHoistable(instance);
+    const linkInstance: HTMLLinkElement = (instance: any);
+    (linkInstance: any)._p = new Promise((resolve, reject) => {
+      linkInstance.onload = resolve;
+      linkInstance.onerror = reject;
+    });
+    setInitialProperties(instance, 'link', stylesheetProps);
+    resource.instance = instance;
+    state.stylesheets.set(resource, hoistableRoot);
+
+    const preloadEl = resource.state.preload;
+    if (preloadEl && (resource.state.loading & Settled) === NotLoaded) {
+      state.count++;
+      state.preloadCount++;
+      const ping = onPreloadComplete.bind(state);
+      preloadEl.addEventListener('load', ping);
+      preloadEl.addEventListener('error', ping);
+    }
+  }
+}
+
+export function waitForCommitToBeReady(): null | (Function => Function) {
+  if (suspendedState === null) {
+    throw new Error(
+      'Internal React Error: suspendedState null when it was expected to exists. Please report this as a React bug.',
+    );
+  }
+
+  const state = suspendedState;
+
+  if (state.stylesheets.size > 0 && state.count === 0) {
+    // We are not currently blocked but we have not inserted all stylesheets.
+    insertSuspendedStylesheets(state, state.stylesheets);
+  }
+
+  // We need to check the count again because the inserted stylesheets may have led to new
+  // tasks to wait on.
+  if (state.count > 0) {
+    return commit => {
+      state.unsuspend = commit;
+      // In theory we don't need to bind this because we should always cancel
+      // before starting a new suspendedState
+      return cancelSuspend;
+    };
+  }
+  return null;
+}
+
+function cancelSuspend() {
+  // We expect cancelSuspend to be called before calling `startSuspendingCommit` againt.
+  // If this constraint is violated this will not work right and will need to be bound instead
+  const state: SuspendedState = (suspendedState: any);
```

Comment:
If there are multiple roots, they are allowed to commit concurrently. You can only treat `suspendedState` as atomic until `waitForCommitToBeReady` is called.

Reply:
yeah good call I missed that